next/frontend/utils/queryParamRedirect.ts

```diff
@@ -0,0 +1,151 @@
+import qs from 'qs'
+
+import { ROUTES } from '../api/constants'
+import { isProductionDeployment } from './general'
+
+const productionOrigins = ['https://kupaliska.bratislava.sk', 'https://bratislava.sk']
+const stagingOrigins = [
+  'http://localhost:3000',
+  'http://localhost:3001',
+  'http://localhost:3002',
+  'http://localhost:3003',
+  'https://kupaliska.staging.bratislava.sk',
+  'https://kupaliska.dev.bratislava.sk',
+]
+
+export const APPROVED_LOGIN_REDIRECT_ORIGINS = isProductionDeployment()
+  ? productionOrigins
+  : [...productionOrigins, ...stagingOrigins]
+
+export enum SafeRedirectType {
+  Local = 'local',
+  Remote = 'remote',
+}
+
+export const redirectQueryParam = 'from'
+
+export type SafeRedirect = {
+  url: string
+  type: SafeRedirectType
+}
+
+const defaultSafeRedirect: SafeRedirect = {
+  url: ROUTES.HOME,
+  type: SafeRedirectType.Local,
+}
+
+export const isDefaultRedirect = (loginRedirectResult: SafeRedirect) =>
+  loginRedirectResult.url === ROUTES.HOME
+
+/**
+ * Returns a safe variant (including type) of the redirect query param. For remote URLs, we only allow a subset of origins.
+ */
+export const getSafeRedirect = (queryParam: unknown) => {
+  if (typeof queryParam !== 'string') {
+    return defaultSafeRedirect
+  }
+
+  if (queryParam.startsWith('/')) {
+    return {
+      url: queryParam,
+      type: SafeRedirectType.Local,
+    }
+  }
+
+  try {
+    const url = new URL(queryParam)
+    if (APPROVED_LOGIN_REDIRECT_ORIGINS.includes(url.origin)) {
+      return {
+        url: queryParam,
+        type: SafeRedirectType.Remote,
+      }
+    }
+    // eslint-disable-next-line no-empty
+  } catch (error) {}
+
+  return defaultSafeRedirect
+}
+
+/**
+ * We want to keep only viable and safe query params in the URL, otherwise it is removed.
+ */
+export const shouldRemoveRedirectQueryParam = (originalQueryParam: unknown) => {
+  if (typeof originalQueryParam !== 'string') {
+    return false
+  }
+
+  const safeRedirect = getSafeRedirect(originalQueryParam)
+  if (isDefaultRedirect(safeRedirect)) {
+    return true
+  }
+
+  // If the "safe redirect" is different from the original query param, we want to remove the original query param
+  return safeRedirect?.url !== originalQueryParam
+}
+
+/**
+ * Parses the resolved URL and removes the redirect query param if it exists.
+ *
+ * @example
+ * /some-path?from=/some-other-path -> /some-path
+ * /some-path?from=/some-other-path&other=param -> /some-path?other=param
+ * /some-path -> /some-path
+ */
+export const removeRedirectQueryParamFromUrl = (resolvedUrl: string) => {
+  const splitUrl = resolvedUrl.split('?')
+  if (splitUrl.length === 1) {
+    return resolvedUrl
+  }
+
+  const [url, ...queryStringArray] = splitUrl
+  const queryString = queryStringArray.join('')
+
+  const queryParams = qs.parse(queryString)
```

Comment:
does this work well with hash at the end of the url ? if not, even if it's not used in context with hashes today, can it be updated so that it's not a problem ?

Reply:
Hashes are handled completely client side (they are not passed to the server, I didn't know either).